analytics/src/main/java/com/segment/analytics/Underwriter.java

```diff
@@ -0,0 +1,41 @@
+package com.segment.analytics;
+
+import android.util.*;
+
+import java.security.InvalidKeyException;
+import java.security.NoSuchAlgorithmException;
+
+import javax.crypto.Mac;
+import javax.crypto.spec.SecretKeySpec;
+
+public class Underwriter {
+    private Mac mac;
+
+    public Underwriter(String key) {
+        if (key == null) {
+            key = "segment";
+        }
+
+        try {
+            String type = "HmacSHA1";
+            mac = Mac.getInstance(type);
+            SecretKeySpec secret = new SecretKeySpec(key.getBytes(), type);
+            mac.init(secret);
+        } catch (NoSuchAlgorithmException | InvalidKeyException e) {
+            e.printStackTrace();
```

Comment:
I think the `NoSuchAlgorithmException` is pretty unlikely, but what about `InvalidKeyException`?  IIUC, if we return here and the `mac` is empty, then we'd get NPEs later when trying to send the batch.  Would we want to either 1) throw out an exception so the caller can fall back to not signing or 2) log a warning and check `mac` for `null` in the `update` and `sign` methods and fall back to no signing?

Reply:
that's a great idea. we have this implemented later